packages/hoppscotch-data/src/rest/index.ts

```diff
@@ -1,66 +1,53 @@
-import cloneDeep from "lodash/cloneDeep"
-import * as Eq from "fp-ts/Eq"
-import * as S from "fp-ts/string"
-import { ValidContentTypes } from "./content-types"
-import { HoppRESTAuth } from "./HoppRESTAuth"
-import { lodashIsEqualEq, mapThenEq, undefinedEq } from "../utils/eq"
+import * as Eq from "fp-ts/Eq";
+import * as S from "fp-ts/string";
+import cloneDeep from "lodash/cloneDeep";
+import { createVersionedEntity, InferredEntity } from "verzod";
 
-export * from "./content-types"
-export * from "./HoppRESTAuth"
-
-export const RESTReqSchemaVersion = "1"
-
-export type HoppRESTParam = {
-  key: string
-  value: string
-  active: boolean
-}
-
-export type HoppRESTHeader = {
-  key: string
-  value: string
-  active: boolean
-}
+import { lodashIsEqualEq, mapThenEq, undefinedEq } from "../utils/eq";
+import V0_VERSION from "./v/0";
+import V1_VERSION, { HoppRESTAuth, HoppRESTHeaders, HoppRESTParams, HoppRESTReqBody } from "./v/1";
 
-export type FormDataKeyValue = {
-  key: string
-  active: boolean
-} & ({ isFile: true; value: Blob[] } | { isFile: false; value: string })
-
-export type HoppRESTReqBodyFormData = {
-  contentType: "multipart/form-data"
-  body: FormDataKeyValue[]
-}
-
-export type HoppRESTReqBody =
-  | {
-      contentType: Exclude<ValidContentTypes, "multipart/form-data">
-      body: string
-    }
-  | HoppRESTReqBodyFormData
-  | {
-      contentType: null
-      body: null
+export * from "./content-types"
+export {
+  FormDataKeyValue,
+  HoppRESTReqBodyFormData,
+  HoppRESTAuth,
+  HoppRESTAuthAPIKey,
+  HoppRESTAuthBasic,
+  HoppRESTAuthBearer,
+  HoppRESTAuthNone,
+  HoppRESTAuthOAuth2,
+  HoppRESTReqBody
+} from "./v/1"
+
+export const HoppRESTRequest = createVersionedEntity({
+  latestVersion: 1,
+  versionMap: {
+    0: V0_VERSION,
+    1: V1_VERSION
+  },
+  getVersion(data) {
+    // For V1 onwards we have the v string storing the number
+    if (
+      typeof data === "object"
+      && data !== null
+      && "v" in data
+      && typeof data.v === "string"
+      && !Number.isNaN(parseInt(data.v))
+    ) {
+      return parseInt(data.v)
     }
 
-export interface HoppRESTRequest {
-  v: string
-  id?: string // Firebase Firestore ID
+    // For V0 we have to check the schema
+    const result = V0_VERSION.schema.safeParse(data)
 
-  name: string
-  method: string
-  endpoint: string
-  params: HoppRESTParam[]
-  headers: HoppRESTHeader[]
-  preRequestScript: string
-  testScript: string
+    return result.success ? 0 : null
+  },
```

Comment:
If we want we could write this in a zod native way. but it'll feel a little verbose. something along the lines of,

```typescript
getVersion(data) {
    const v1AndAbove = z.object({
      // For V1 onwards we have the v string storing the number
      v: z.string().transform((versionString, ctx) => {
        const v = parseInt(versionString)

        if (isNaN(v)) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: "Cannot parse valid number from version string",
          })

          return z.NEVER
        }

        return v
      }),
    })

    const allVersionsSchema = z.union([
      v1AndAbove,
      // for v0, we don't have a v key, so we have to check if the schema matches
      V0_VERSION.schema.transform((schema) => ({
        v: 0,
      })),
    ])

    const result = allVersionsSchema.safeParse(data)
    return result.success ? result.data.v : null
  },
```